enableWhenExpression follow up PR for maintaining CodeHealth

RaaziaTarique opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1345 

**Description**
Remove questionnaireResponseItemRetriever and instead implement the functionality of finding the questionnaire response item from linkId inside of the enablement package.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Code health

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [ ] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

maimoonak commented on datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:
```diff
@@ -164,3 +150,25 @@ val fhirPathEngine: FHIRPathEngine =
   }
+
+/** Map from link IDs to questionnaire response items. */
+private fun linkIdToQuestionnaireResponseItemMap(questionnaireResponse: QuestionnaireResponse) =
```
there is a similar method/map in QuestionnaireViewModel (linkIdToQuestionnaireResponseItemMap). We should utilize that. This map creation every time for each evaluation would be expensive. Or is there any specific reason to not use that map!

RaaziaTarique replied:
We were previously using the same mapping but it was being done in the QuestionnaireViewModel as we were not passing the QuestionnaireResponse to EnablementEvaluator. I can use the same method present QuestionnaireViewModel though.

jingtang10 commented:
@joiskash please take a look at this. We spoke about this before. This is a really important refactoring.

joiskash reviewed (commented):
These changes look good, but I am unable to figure out one part:

In `questionnaireResponseItemChangedCallback` the `linkIdToQuestionnaireResponseItemMap` is modified. Is this map equivalent to the map that you get after `createLinkIdToQuestionnaireResponseItemMap` with the updated questionnaireResponse ?

RaaziaTarique commented:
> These changes look good, but I am unable to figure out one part:
> 
> In `questionnaireResponseItemChangedCallback` the `linkIdToQuestionnaireResponseItemMap` is modified. Is this map equivalent to the map that you get after `createLinkIdToQuestionnaireResponseItemMap` with the updated questionnaireResponse ?
> 
> @RaaziaTarique @jingtang10

Yes, as I have moved createLinkIdToQuestionnaireResponseItemMap to `MoreQuestionnaireItemComponent` so, I modified it as per  `EnablementEvaluator`.
@joiskash   CC: @jingtang10

Tarun-Bhardwaj commented:
@f-odhiambo are you aware if this PR is ready for review by Google?

f-odhiambo commented:
> @f-odhiambo are you aware if this PR is ready for review by Google?

This PR is ready for review @Tarun-Bhardwaj

jingtang10 commented:
i'll raise a separate pr
